manager/src/main/java/org/openremote/manager/energy/GOPACSHandler.java

```diff
@@ -0,0 +1,481 @@
+                uftpSendMessageService.attemptToSendMessage(
+                        message.payloadMessage(),
+                        new SigningDetails(
+                                message.sender(),
+                                this.privateKey,
+                                new UftpParticipant(recipientDomain, UftpRoleInformation.getRecipientRoleBySenderRole(message.sender().role()))
+                        )
+                );
+            }
+            LOG.fine("Finished sending message: " + message);
+        } catch (Exception e) {
+            LOG.log(Level.SEVERE, "Failed to send message", e);
+        }
+    }
+
+    @Override
+    public String getAuthorizationHeader(UftpParticipant uftpParticipant) {
+        return "";
+    }
+
+    protected void handleFlexRequestMessage(UftpParticipant participant, FlexRequest flexRequest) {
+        LOG.fine("Received Flex Request: " + flexRequest);
+        int year = flexRequest.getPeriod().getYear();
+        int month = flexRequest.getPeriod().getMonthValue();
+        int day = flexRequest.getPeriod().getDayOfMonth();
+        List<ValueDatapoint<?>> powerImportMaxGopacsDatapoints = new ArrayList<>();
+        List<ValueDatapoint<?>> powerExportMaxGopacsDatapoints = new ArrayList<>();
+
+        for (int i = 0; i < flexRequest.getISPS().size(); i++) {
+            FlexRequestISPType flexRequestISPType = flexRequest.getISPS().get(i);
+            LocalTime start = FlexRequestISPTypeHelper.getISPStart(flexRequestISPType.getStart(), year, month, day, flexRequest.getTimeZone());
+            double importMax = flexRequestISPType.getMaxPower() == 0L ? this.powerImportMax : (double) flexRequestISPType.getMaxPower() / 1000.0F;
+            double exportMax = flexRequestISPType.getMinPower() == 0L ? this.powerExportMax : (double) Math.abs(flexRequestISPType.getMinPower()) / 1000.0F;
+            this.schedulePowerUpdate(start, ElectricitySupplierAsset.POWER_IMPORT_MAX_GOPACS.getName(), exportMax);
+            this.schedulePowerUpdate(start, ElectricitySupplierAsset.POWER_EXPORT_MAX_GOPACS.getName(), exportMax);
+
+            // Correct usage of ZoneId.of instead of ZoneOffset.of
+            ZoneId zoneId = ZoneId.of(flexRequest.getTimeZone());
+            long startEpochMilli = flexRequest.getPeriod().atTime(start).atZone(zoneId).toInstant().toEpochMilli();
+
+            powerImportMaxGopacsDatapoints.add(new ValueDatapoint<>(startEpochMilli, importMax));
+            powerExportMaxGopacsDatapoints.add(new ValueDatapoint<>(startEpochMilli, exportMax));
+        }
+
+        this.setPredictedDataPoints(ElectricitySupplierAsset.POWER_IMPORT_MAX_GOPACS.getName(), powerImportMaxGopacsDatapoints);
+        this.setPredictedDataPoints(ElectricitySupplierAsset.POWER_EXPORT_MAX_GOPACS.getName(), powerExportMaxGopacsDatapoints);
+        LOG.fine("Finished processing Flex Request: " + flexRequest);
+    }
```

Comment:
I don't get why this GOPACS handler is directly handling the request messages and then updating attribute values. I was expecting this handler to just route requests and send responses, energy optimisation service should do the decision making and updating attributes. e.g. GOPACSHandler just exposes methods/callbacks for accepting/confirming requests.

Reply:
So this is just an initial implementation? In early discussions we discussed creating an interface that energy optimisation can be written against. Not clear to me why we would want to merge this in its' current form.